core/src/layers/correctness_check.rs

```diff
@@ -0,0 +1,602 @@
+    RpDelete, RpList, RpPresign, RpRead, RpRename, RpStat, RpWrite,
+};
+use crate::{Error, ErrorKind};
+
+/// Add a correctness capability check layer for every operation
+///
+/// Before performing any operations, we will first verify the operation and its critical arguments
+/// against the capability of the underlying service. If the operation or arguments is not supported,
+/// an error will be returned directly.
+///
+/// # Notes
+///
+/// OpenDAL applies this checker to every accessor by default, so users don't need to invoke it manually.
+/// this checker ensures the operation and its critical arguments, which might affect the correctness of
+/// the call, are supported by the underlying service.
+///
+/// for example, when calling `write_with_append`, but `append` is not supported by the underlying
+/// service, an `Unsupported` error is returned. without this check, undesired data may be written.
+#[derive(Default)]
+pub struct CorrectnessChecker;
+
+impl<A: Access> Layer<A> for CorrectnessChecker {
+    type LayeredAccess = CorrectnessAccessor<A>;
+
+    fn layer(&self, inner: A) -> Self::LayeredAccess {
+        CorrectnessAccessor {
+            info: inner.info(),
+            inner,
+        }
+    }
+}
+
+pub(crate) fn new_unsupported_error(info: &AccessorInfo, op: impl Into<&'static str>) -> Error {
+    let scheme = info.scheme();
+    let op = op.into();
+
+    Error::new(
+        ErrorKind::Unsupported,
+        format!("service {scheme} doesn't support operation {op}"),
+    )
+    .with_operation(op)
+}
+
+pub(crate) fn new_unsupported_args_error(
+    info: &AccessorInfo,
+    op: impl Into<&'static str>,
+    args: &str,
+) -> Error {
+    let scheme = info.scheme();
+    let op = op.into();
+
+    Error::new(
+        ErrorKind::Unsupported,
+        format!("service {scheme} doesn't support operation {op} with args {args}"),
+    )
+    .with_operation(op)
+}
+
+pub struct CorrectnessAccessor<A: Access> {
+    info: Arc<AccessorInfo>,
+    inner: A,
+}
+
+impl<A: Access> Debug for CorrectnessAccessor<A> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("CorrectnessCheckAccessor")
+            .field("inner", &self.inner)
+            .finish_non_exhaustive()
+    }
+}
+
+impl<A: Access> LayeredAccess for CorrectnessAccessor<A> {
+    type Inner = A;
+    type Reader = A::Reader;
+    type BlockingReader = A::BlockingReader;
+    type Writer = A::Writer;
+    type BlockingWriter = A::BlockingWriter;
+    type Lister = A::Lister;
+    type BlockingLister = A::BlockingLister;
+
+    fn inner(&self) -> &Self::Inner {
+        &self.inner
+    }
+
+    fn info(&self) -> Arc<AccessorInfo> {
+        self.info.clone()
+    }
+
+    async fn create_dir(&self, path: &str, args: OpCreateDir) -> crate::Result<RpCreateDir> {
+        let capability = self.info.full_capability();
+        if !capability.create_dir {
```

Comment:
This check is unnecessary as we will fall back to the default implementation of `Access`.

Reply:
Ummm, so it seems we don't need to check `operation` itself ?

Reply:
Yes, we don't need to check operation itself. All of them are already covered.